packages/@react-spectrum/list/docs/ListView.mdx

```diff
@@ -0,0 +1,941 @@
+            <Text>{item.name}</Text>
+            {item.type === 'folder' &&
+              <>
+                <Folder />
+                <Text slot="description">{`contains ${item.childNodes.length} dropped item(s)`}</Text>
+              </>
+            }
+          </Item>
+        )}
+      </ListView>
+    </Flex>
+  );
+}
+```
+</details>
+
+```tsx example
+<DragIntoListFilesOnly />
+```
+
+The example below demonstrates how to make a ListView draggable and droppable at the same time. The ListView below supports reordering its own rows via drag and drop.
+<details>
+  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>
+
+```tsx example export=true render=false
+function ReorderableList() {
+  let list = useListData({
+    initialItems: [
+      {id: '1', type: 'file', name: 'Adobe Photoshop'},
+      {id: '2', type: 'file', name: 'Adobe XD'},
+      {id: '3', type: 'folder', name: 'Documents', childNodes: []},
+      {id: '4', type: 'file', name: 'Adobe InDesign'},
+      {id: '5', type: 'folder', name: 'Utilities', childNodes: []},
+      {id: '6', type: 'file', name: 'Adobe AfterEffects'}
+    ]
+  });
+
+  // Append a generated key to the item type so they can only be reordered within this list and not dragged elsewhere.
+  let dragType = React.useMemo(() => `keys-${Math.random().toString(36).slice(2)}`, []);
+  let acceptedDragTypes = [`${dragType}-folder`, `${dragType}-file`];
+
+  // Update the list order based on the position of the drop and the set of rows that were dropped
+  let onMove = (keys, target) => {
+    switch (target.dropPosition) {
+      case 'before':
+        list.moveBefore(target.key, keys);
+        break;
+      case 'after':
+        list.moveAfter(target.key, keys);
+        break;
+      case 'on': {
+        let targetFolder = list.getItem(target.key);
+        let draggedItems = keys.map((key) => list.getItem(key));
+        list.update(target.key, {...targetFolder, childNodes: [...targetFolder.childNodes, ...draggedItems]});
+        list.remove(...keys);
+        break;
+      }
+    }
+  };
+
+  let dragHooks = useDragHooks({
+    getItems(keys) {
+      return [...keys].map(key => {
+        let item = list.getItem(key);
+        // Setup the info for each item that will be propagated upon drag
+        return {
+          [`${dragType}-${item.type}`]: JSON.stringify(item)
+        };
+      });
+    }
+  });
+
+  let dropHooks = useDropHooks({
+    async onDrop(e) {
+      let keysToMove = [];
+      // Only add the dropped item to the list of keys to move
+      // if the dropped item has the expected types
+      for (let item of e.items) {
+        for (let acceptedType of acceptedDragTypes) {
+          if (item.kind === 'text' && item.types.has(acceptedType)) {
+            let {id} = JSON.parse(await item.getText(acceptedType));
+            keysToMove.push(id);
+          }
+        }
+      }
```

Comment:
Ideally I would've liked to exit the inner loop early upon finding the matching `acceptedType` but `e.items` can either be an array of length X, where X is the number of dragged items OR is an array of a single item with a `types` key that is a Set of size X. The second case happens specifically if dragging multiple items that have different types via mouse.

Still digging into why this is and whether or not it is expected

Reply:
Looks to be due to https://github.com/adobe/react-spectrum/blob/26990c2dc4e908ce939d0cb2e10ce5c1cbfe579e/packages/%40react-aria/dnd/src/utils.ts#L111-L126

When there are multiple items w/ the same type or a item with multiple types present in the drag operation, we enter the custom data path, but if you drag a single folder + single item then we hit this path:
https://github.com/adobe/react-spectrum/blob/26990c2dc4e908ce939d0cb2e10ce5c1cbfe579e/packages/%40react-aria/dnd/src/utils.ts#L117-L120
resulting in a `DataTransferItemList` of two `DataTransferItems`

This will then hit the non-custom data path in `readFromDataTransfer`, eventually cumulating to 
https://github.com/adobe/react-spectrum/blob/26990c2dc4e908ce939d0cb2e10ce5c1cbfe579e/packages/%40react-aria/dnd/src/utils.ts#L231-L238
which will transform a `stringItems` list of
```
stringItems Map(2) {'file' => 'Adobe Photoshop', 'folder' => 'Documents'}
```
to a single item representation of 
<img width="424" alt="image" src="https://user-images.githubusercontent.com/9661127/172261504-fea45f4f-dafe-4e6a-8f07-dbc7046faf2a.png">

This feels like it is intentional due to limitations in the current DataTransferAPI, but the comment "All string items are different representations of the same item" sounds incorrect here? Not entirely sure what that means tbh, need to dig more into the DataTransferApi

Reply:
All in all, it feels hard to tell between a drag + drop operation of a single file + folder and a drag + drop operation of a single item that has multiple types tied to it